File: Ad-Hoc/Python/URI-1246.py
def main():
    while True:
        try:
            size, operations = [int(number) for number in input().split()]

            parking = "-" * size
            parking_dict = {}
            balance = 0

            for i in range(operations):
                entry = input().split()

                if len(entry) == 3:
                    status, car_palate, car_size = entry[0], entry[1], entry[2]

                elif len(entry) == 2:
                    status, car_palate = entry[0], entry[1]

                if status == "C":  # Verify possible space to park
                    handle = parking.find("-" * int(car_size))

                    if handle != -1:
                        parking = parking.replace(
                            "-" * int(car_size), "*" * int(car_size), 1
                        )

                        parking_dict[car_palate] = (
                            handle,
                            handle + int(car_size) - 1,
                        )

                        balance += 10

                else:
                    initial = parking_dict[car_palate][0]
                    final = parking_dict[car_palate][1]

                    if final + 1 == len(parking):  # Adjusting the right limit
                        final = len(parking) - 1

                    parking = ("-" * (final - initial + 1)).join(
                        [parking[0:initial], parking[final + 1 : len(parking)]]
                    )

                    parking_dict.pop(car_palate)

            print(balance)

        except EOFError:  # Change for EOFError when running in Ideone
            break



main()

File: Ad-Hoc/Java/URI-1246.java
import java.io.BufferedReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		while(true) {
			String readLine = br.readLine();
			
			if(readLine == null) break;
			
			String[] entry = readLine.split(" ");
			
			int size = Integer.valueOf(entry[0]);
			int operations = Integer.valueOf(entry[1]);
			
			Parking parking = new Parking(size);		
			Map<String, Integer[]> parkingMap = new HashMap<String, Integer[]>();
			
			int balance = 0;
			
			for(int i = 0; i < operations; i++) {
				String[] data = br.readLine().split(" ");
				
				String status;
				String carPalate;
				int carSize;
				
				if(data.length == 3) {
					
					status = data[0];
					carPalate = data[1];
					carSize = Integer.parseInt(data[2]);
					
					StringBuilder strBuilder = new StringBuilder();
					for(int j = 0; j < carSize; j++) {
						strBuilder.append('-');
					}
					
					String parkSpace = new String(strBuilder);					
					int beginning = parking.getSize().indexOf(parkSpace);
					
					if(beginning != -1) {
						parking.occupyParking(parkSpace, beginning);
						
						parkingMap.put(carPalate, new Integer[] {beginning, beginning + carSize});
						
						balance += 10;
						
					}				
					
				}
				
				else{
					status = data[0];
					carPalate = data[1];
					
					int initial = parkingMap.get(carPalate)[0];
					int end = parkingMap.get(carPalate)[1];
					
					parking.vacateSpace(initial, end);
					
					parkingMap.remove(carPalate);
					
				}
				
			}
			
			System.out.println(balance);
			
			
		}

	}
	
	static class Parking{
		
		private StringBuilder size = new StringBuilder();	
		
		public Parking(int spaces) {
			for (int i = 0; i < spaces; i++) {
				size.append('-');
				
			}
			
		}
		
		public StringBuilder getSize() {
			
			return size;
		}
		
		public void occupyParking(String spaceToPark, int beginning) {
			
			StringBuilder toPark = new StringBuilder();
			
			for(int i = 0; i < spaceToPark.length(); i++) {
				toPark.append("*");		
			}
			
			String car = new String(toPark);
			
			size.replace(beginning, beginning + car.length(), car);
			
		}
		
		public void vacateSpace(int beginning, int end) {
			
			int carSize = end - beginning;
			
			StringBuilder toRelease = new StringBuilder();
			
			for(int i = 0; i < carSize; i++) {
				toRelease.append("-");		
			}
			
			String freeSpace = new String(toRelease);
			
			size.replace(beginning, end, freeSpace);
			
		}
		
	}
}
